### DISCUSSION TITLE
Use inputs vars inside an expression function like hashFiles

### DISCUSSION THREAD

**Discussion by tomgrossman** [community/community#25761] [Actions]
<p>I would like to use the <code>hashFiles</code> function dynamically. In short - passing it the <code>key</code> and <code>path</code> as <code>inputs</code>.<br>
We have multiple workflows, for each service/deployment, each with it’s own <code>package.json</code> in its own path in a monolith.<br>
<strong>Even in one workflow, there are multiple jobs that require this same steps.</strong><br>
In order not to duplicate too much code, we tried to create a <code>composite</code> step action for the node setup and the cache usage.<br>
This is how the composite action (should) looks like:</p>
<pre><code class="lang-auto">name: 'Setup node environment'
description: Install correct version of node and download cache
inputs:
  key:
    required: true
    description: cache key
  path:
    required: true
    description: cache path
outputs:
  cache-hit:
    description: cache-hit outputs from actions/cache@v2
    value: ${{ steps.cache.outputs.cache-hit }}
runs:
  using: "composite"
  steps:
    - uses: actions/setup-node@v2
      with:
        node-version: 16
    - uses: actions/cache@v2
      id: cache
      with:
        path: ${{ inputs.path }}
        key: ${{ hashFiles(${{ inputs.key }}) }}
</code></pre>
<p>However, using this action is resulting with the following error:</p>
<pre><code class="lang-auto">Error: GitHub.DistributedTask.ObjectTemplating.TemplateValidationException: The template is not valid. /home/runner/work/dashboard/dashboard/./.github/actions/setup-node/action.yml (Line: 21, Col: 14): Unexpected symbol: '${{'. Located at position 11 within expression: hashFiles(${{ inputs.key
   at GitHub.DistributedTask.ObjectTemplating.TemplateValidationErrors.Check()
   at GitHub.Runner.Worker.ActionManifestManager.ConvertRuns(IExecutionContext executionContext, TemplateContext templateContext, TemplateToken inputsToken, String fileRelativePath, MappingToken outputs)
   at GitHub.Runner.Worker.ActionManifestManager.Load(IExecutionContext executionContext, String manifestFile)
</code></pre>
<p>I looked in the community and I found a similar <a href="https://github.community/t/is-hashfiles-of-actions-cache-can-be-dynamic/171342">issue</a> but with an insufficient solution since it is using <code>context</code> values (like <code>github.workspace</code>, <code>matrix.rails-version</code>) and not inputs values like I need.</p>
<p>So, is there a solution for this?</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="tomgrossman" data-post="1" data-topic="221148">
<div class="title">
<div class="quote-controls"></div>
 tomgrossman:</div>
<blockquote>
<p><code>        key: ${{ hashFiles(${{ inputs.key }}) }}</code></p>
</blockquote>
</aside>
<p>You have a template wrapped in template there, try:</p>
<pre><code class="lang-auto">        key: ${{ hashFiles(inputs.key) }}
</code></pre>

**Comment by tomgrossman** [comment]
<p>OK wow, how did I now tried that? I though inputs must be accessed explicitly with `${{ }}</p>

**Comment by airtower-luna** [comment]
<p>It’s not inputs, the Actions expressions need to be wrapped in those. Within an expression you can use variables by name. :smiley_cat:</p>

**Comment by ErcinDedeoglu** [comment]
Upvoted. Thank you for your answer. My case is different:

this is what I have:
```
key: ${{ matrix.projectName }}-Library-${{ hashFiles('./repo/Alpha/**', './repo/Beta/**', './repo/Zulu/**') }}
```
what I need:


```
   strategy:
      matrix:
        path:
          - ./repo
         
key: ${{ matrix.projectName }}-Library-${{ hashFiles('${{ matrix.repoLocalPath }}/Alpha/**', '${{ matrix.repoLocalPath }}/Beta/**', '${{ matrix.repoLocalPath }}/Zulu/**') }}
```


I want to change the hardcoded "./repo" string with ${{ matrix.repoLocalPath }} and concat it with a Hardcoded tail.

Any solution for this?

**Reply by airtower-luna** [reply]
Try using [`format()`](https://docs.github.com/en/actions/learn-github-actions/expressions#format) to build the inner string, like so:

```
${{ hashFiles(format('{0}/Alpha/**', matrix.repoLocalPath)) }}
```

**Reply by ErcinDedeoglu** [reply]
thanks :)

**Reply by joshjohanning** [reply]
nice that's super helpful. Mine looks like:

```yml
key: ${{ runner.os }}-${{ hashFiles(format('{0}/{1}',inputs.folder-to-hash,'package.json')) }}
```

**Reply by brianbruggeman** [reply]
Does this actually work?  I think on my system running `act` it ends up looking like:

```
key: ${{ runner.os }}-
```

Without an error.

**Reply by joshjohanning** [reply]
@brianbruggeman my snippet ☝️  works for me. Maybe `act` is unable to process this the same as Actions

Code:
- https://github.com/joshjohanning-org/expressions-testing/blob/main/.github/workflows/reusable-workflow.yml#L38

Log line in Action run:
- https://github.com/joshjohanning-org/expressions-testing/actions/runs/7980811233/job/21791242710#step:5:56

**Comment by PTruscott** [comment]
Sorry to resurrect an old thread - I have a similar issue.  I have an arbitrary number of file paths I'd like to use in `hashFiles`.
I have a string variable `FOLDER_PATHS` which is the output of a previous action.  

```
echo ${{steps.get-hash-paths.outputs.FOLDER_PATHS}} # this outputs what I expect
echo ${{ hashFiles(steps.get-hash-paths.outputs.FOLDER_PATHS) }} # this outputs absolutely nothing, no error
echo ${{ hashFiles('./src', './src/analytics') }} # when manually copying in the output from the first step this produces a hash as desired.
```
I've also tried using a bash array, but trying to use bash variables with the `$` syntax leads to syntax errors.
Anyone know the solution to get this working?

**Reply by copdips** [reply]
would be better if hashFiles accepts syntax like `echo ${{ hashFiles('./src,./src/analytics') }}`

**Reply by berleant** [reply]
As far as I can tell, there's just no way to do this.

**Comment by brun0xff** [comment]
Taking advantage of the discussion, I think the hashFiles function only works with hardcoded parameters, I tried something like:
```
echo ${{ hashFiles($FOO) }}
```
And it works as if the contents of $FOO were empty, in the same step I echo the variable and see its contents

**Reply by gionn** [reply]
$FOO it's a shell notation and can't work inside `${{ }}`, you should use env.FOO inside `${{ }}`
